### PR TITLE
optimize the query search

### PR DESCRIPTION
**Pull request by Omarabdul3ziz**
### Changes
- apply case insensitive search
- and the substring search 

on the following queries
  - `nodes?country`
  - `nodes?city`
  - `nodes?farm_name`
  - `gateways?country`
  - `gateways?city`
  - `gateways?farm_name`
  - `farms?name`

### Related issues
- https://github.com/threefoldtech/tfgridclient_proxy/issues/175

### CODE REVIEW

**Comment by Omarabdul3ziz** [comment]
After a discussion with Hanafy we better has a separate query for the substring match and another for the literal match. so it should be: 
- `nodes?[farm_name=, city=, country= ]` and `farms?name=` : literal matching with case insensitive.
- `nodes?[farm_name_contains=, city_contains=, country_contains= ]` and `farms?name_contains=` : substring matching.


Work on that still in-progress.

File: internal/explorer/db/postgres.go [hunk 1]
```diff
@@ -379,10 +379,16 @@ func (d *PostgresDatabase) GetNodes(filter types.NodeFilter, limit types.Limit)
 		q = q.Where("nodes_resources_view.free_sru >= ?", *filter.FreeSRU)
 	}
 	if filter.Country != nil {
-		q = q.Where("node.country = ?", *filter.Country)
+		q = q.Where("LOWER(node.country) = LOWER(?)", *filter.Country)
+	}
+	if filter.CountryContains != nil {
+		q = q.Where("LOWER(node.country) LIKE '%' || LOWER(?) || '%'", *filter.CountryContains)
```
Comment (AhmedHanafy725) [inline internal/explorer/db/postgres.go]: why not use `ILIKE` instead?

File: internal/explorer/db/postgres.go [hunk 2]
```diff
@@ -379,10 +379,16 @@ func (d *PostgresDatabase) GetNodes(filter types.NodeFilter, limit types.Limit)
...
 	}
 	if filter.City != nil {
-		q = q.Where("node.city = ?", *filter.City)
+		q = q.Where("LOWER(node.city) = LOWER(?)", *filter.City)
+	}
+	if filter.CityContains != nil {
+		q = q.Where("LOWER(node.city) LIKE '%' || LOWER(?) || '%'", *filter.CityContains)
```
Comment (AhmedHanafy725) [inline internal/explorer/db/postgres.go]: same here

File: internal/explorer/db/postgres.go [hunk 3]
```diff
@@ -394,7 +400,10 @@ func (d *PostgresDatabase) GetNodes(filter types.NodeFilter, limit types.Limit)
 		q = q.Where("node.farm_id IN ?", filter.FarmIDs)
 	}
 	if filter.FarmName != nil {
-		q = q.Where("farm.name = ?", *filter.FarmName)
+		q = q.Where("LOWER(farm.name) = LOWER(?)", *filter.FarmName)
+	}
+	if filter.FarmNameContains != nil {
```
Comment (AhmedHanafy725) [inline internal/explorer/db/postgres.go:396]: same here

**Review by AhmedHanafy725 — APPROVED** [review]

**Review by AhmedHanafy725 — APPROVED** [review]

**Pull request merged**